programs/drift/src/state/perp_market.rs

```diff
@@ -860,6 +849,80 @@ impl PerpMarket {
             oracle_price_data,
         )?)
     }
+
+    pub fn amm_can_fill_order(
+        &self,
+        order: &Order,
+        clock_slot: u64,
+        fill_mode: FillMode,
+        state: &State,
+        safe_oracle_validity: OracleValidity,
+        user_can_skip_auction_duration: bool,
+        mm_oracle_price_data: &MMOraclePriceData,
+    ) -> DriftResult<bool> {
+        if self.is_operation_paused(PerpOperation::AmmFill) {
+            return Ok(false);
+        }
+
+        if self.has_too_much_drawdown()? {
+            return Ok(false);
+        }
+
+        // We are already using safe oracle data from MM oracle.
+        // But AMM isnt available if we could have used MM oracle but fell back due to price diff
+        // This is basically early volatility protection
+        let mm_oracle_not_too_volatile =
+            if mm_oracle_price_data.is_enabled() && mm_oracle_price_data.is_mm_oracle_as_recent() {
+                let amm_available = !mm_oracle_price_data.is_mm_exchange_diff_bps_high();
+                amm_available
+            } else {
+                true
+            };
+
+        if !mm_oracle_not_too_volatile {
+            msg!("AMM cannot fill order: MM oracle too volatile compared to exchange oracle");
+            return Ok(false);
+        }
+
+        // Determine if order is fillable with low risk
+        let safe_oracle_price_data = mm_oracle_price_data.get_safe_oracle_price_data();
+        let order_is_low_risk_for_amm = order.is_low_risk_for_amm(
+            safe_oracle_price_data.delay,
+            clock_slot,
+            fill_mode.is_liquidation(),
+        )?;
+        let oracle_valid_for_amm_fill_low_risk = is_oracle_valid_for_action(
+            safe_oracle_validity,
+            Some(DriftAction::FillOrderAmmLowRisk),
+        )?;
+        if !oracle_valid_for_amm_fill_low_risk {
```

Comment:
think you can check this before `oracle_valid_for_amm_fill_low_risk` so you save some CUs when it's false.